---
title: "Genome stratification of FP and FN"
output: github_document
---

```{r setup, include=FALSE}
library(here)
library(tidyverse)
library(data.table)
library(GenomicRanges)
library(VariantAnnotation)
library(ggh4x)
source("config.R")
knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center',fig.retina=2)
```

Here, we study the genomic location of FP and FN calls with the genome stratification from GIAB.

```{r cache=TRUE}
cds_bed <- paste0(genome_stratification_dir, "v3.4/GRCh38@all/Functional/GRCh38_refseq_cds.bed.gz")
cds_grange <-  fread(cds_bed, col.names = c("chr","start","end")) |> GRanges()
lc_bed <- paste0(genome_stratification_dir, "v3.4/GRCh38@all/LowComplexity/GRCh38_AllTandemRepeats.bed.gz")
lc_grange <-  fread(lc_bed, col.names = c("chr","start","end")) |> GRanges()
map_bed <- paste0(genome_stratification_dir, "v3.4/GRCh38@all/Mappability/GRCh38_lowmappabilityall.bed.gz")
map_grange <- fread(map_bed, col.names = c("chr","start","end")) |> GRanges()
gc_bed <- paste0(genome_stratification_dir, "v3.4/GRCh38@all/GCcontent/GRCh38_gclt30orgt55_slop50.bed.gz")
gc_grange <- fread(gc_bed, col.names = c("chr","start","end")) |> GRanges()
segdup_bed <- paste0(genome_stratification_dir, "v3.4/GRCh38@all/SegmentalDuplications/GRCh38_segdups.bed.gz")
segdup_grange <- fread(segdup_bed, col.names = c("chr","start","end")) |> GRanges()
```

```{r}
overlap_with_regions <- function(vcf, variant_type="snv") {
  category  <- basename(vcf) |>  str_remove(".vcf")
  tool <- str_extract(vcf, c("clairS","deepsomatic") |> paste(collapse = "|"))
  switch (variant_type,
    snv={variant <- read_tsv(vcf,col_names =FALSE) |> dplyr::select(X1, X2) |>mutate(chr=X1, start=X2, end=X2) |> dplyr::select(chr,start,end) |> dplyr::filter( chr %in% paste0("chr",c(1:22,"X","Y")) ) |> GRanges()},
    indel={variant <- read_tsv(vcf,col_names =FALSE) |> dplyr::select(X1, X2) |> mutate(chr=X1, start=X2-50,end=X2+50) |> dplyr::select(chr,start,end) |> dplyr::filter( chr %in% paste0("chr",c(1:22,"X","Y")) ) |> GRanges()}
  )
  cds <- countOverlaps(variant, cds_grange) |> table() |> as.data.frame() |> mutate(Var1=ifelse(Var1 %in% c("0","1"), as.character(Var1), "1")) |> group_by(Var1) |> summarise(Freq=sum(Freq)) |> ungroup() |> rename(Freq="refseq_cds") 
  lc <- countOverlaps(variant, lc_grange) |> table() |> as.data.frame() |> mutate(Var1=ifelse(Var1 %in% c("0","1"), as.character(Var1), "1")) |> group_by(Var1) |> summarise(Freq=sum(Freq)) |> ungroup() |> rename(Freq="tandem_repeats") 
  map <- countOverlaps(variant, map_grange) |> table() |> as.data.frame() |> mutate(Var1=ifelse(Var1 %in% c("0","1"), as.character(Var1), "1")) |> group_by(Var1) |> summarise(Freq=sum(Freq)) |> ungroup() |> rename(Freq="low_mappability") 
  gc <- countOverlaps(variant, gc_grange) |> table() |> as.data.frame() |> mutate(Var1=ifelse(Var1 %in% c("0","1"), as.character(Var1), "1")) |> group_by(Var1) |> summarise(Freq=sum(Freq)) |> ungroup() |> rename(Freq="GC") 
  segdup <- countOverlaps(variant, segdup_grange) |> table() |> as.data.frame() |> mutate(Var1=ifelse(Var1 %in% c("0","1"), as.character(Var1), "1")) |> group_by(Var1) |> summarise(Freq=sum(Freq)) |> ungroup() |> rename(Freq="segdup") 
  full_join(cds,lc, by="Var1") |> full_join(map) |> full_join(gc) |> full_join(segdup) |> add_column(tool=tool, category=category)
}
```

```{r cache=TRUE, message=FALSE, warning=FALSE}
snv_df <- list.files(paste0(workdir,"analysis/benchmark/snvs/somatic"), pattern = "^fp.vcf|^fn.vcf", full.names = TRUE, recursive = TRUE) |> 
   keep(\(x) grepl("passed",x)) |> 
  set_names(\(f) dirname(f) |> str_split("/") |> map_chr(\(x) last(x))|> str_remove("_passed"))  |> 
  map(overlap_with_regions) |> list_rbind(names_to = "prefix") |> 
  separate_wider_delim(prefix,".",names=c("tumor","tumor_depth","normal","normal_depth")) |> 
  separate_wider_delim(tumor, "_",names=c("cellline","purity"), too_few = "align_start") |> 
  mutate(purity=ifelse(is.na(purity),"100",purity), 
         purity=factor(purity, levels=seq(40,100,20)), 
         Overlap=ifelse(Var1=="1", "Yes","No"),
         comb=paste(tumor_depth,normal_depth,sep="-"))
```

## SNV FP and FN calls genome stratification

```{r fig.height=20, fig.width=9.6}
region_labels <- c("CDS regions","Tandem repeats", "Low mappability","GC% < 30 or >55","Segmental duplications")
names(region_labels) <- c("refseq_cds","tandem_repeats","low_mappability","GC","segdup")
category_labels <- c("False negative", "False positive","True positive")
names(category_labels) <- c("fn","fp","tp")

snv_df |> pivot_longer(c( tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  facet_nested(name + purity~ cellline+category+tool, nest_line = TRUE,
               labeller = labeller(name=region_labels, category=category_labels)) + 
  theme_minimal(base_size = 12) + 
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  labs(x="Proportion", y="Sequencing strategy") + theme(strip.text = element_text(size=12)) + 
  theme(strip.text.y.right = element_text(size=12), axis.text.x.bottom = element_text(size=8))


ggsave("figures/snv_depth_genome_regions.png",height = 20,width = 9.6)
```


## SNV FP and FN calls genome stratification with samples with different purity been merged

```{r, warning=FALSE, message=FALSE,fig.height=7.8,fig.width=9.6}
snv_df |> pivot_longer(c(tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  facet_nested(name ~ cellline+category+tool, nest_line = TRUE,labeller = labeller(name=region_labels, category=category_labels)) + theme_minimal(base_size = 12) +
  labs(x="Proportion", y="Sequencing strategy") + theme(strip.text.y.right = element_text(size=10), axis.text.x.bottom = element_text(size=6))

ggsave("figures/snv_depth_merged_purity_genome_regions.png",height = 7.8,width = 9.6)
```

```{r, warning=FALSE, message=FALSE, cache=TRUE}
indel_df <- list.files(paste0(workdir,"analysis/benchmark/indels/somatic"), pattern = "^fp.vcf|^fn.vcf", full.names = TRUE, recursive = TRUE) |> 
  set_names(\(f) dirname(f) |> str_split("/") |> map_chr(\(x) last(x)))  |> 
  map(\(f) overlap_with_regions(f,variant_type="indel")) |> list_rbind(names_to = "prefix") |> 
  separate_wider_delim(prefix,".",names=c("tumor","tumor_depth","normal","normal_depth")) |> 
  separate_wider_delim(tumor, "_",names=c("cellline","purity"), too_few = "align_start") |> 
  mutate(purity=ifelse(is.na(purity),"100",purity), 
         purity=factor(purity, levels=seq(40,100,20)), 
         Overlap=ifelse(Var1=="1", "Yes","No"),
         comb=paste(tumor_depth,normal_depth,sep="-")) 
```


## INDEL FP and FN calls genome stratification

```{r, warning=FALSE, message=FALSE, fig.height=20, fig.width=9.6}
indel_df |> pivot_longer(c( tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  facet_nested(name + purity~ cellline+category+tool, nest_line = TRUE,
               labeller = labeller(name=region_labels, category=category_labels)) + theme_minimal(base_size = 12) +
  labs( x="Proportion",y="Sequencing strategy") + theme(strip.text.y.right = element_text(size=12), axis.text.x.bottom = element_text(size=6))

ggsave("figures/indel_depth_genome_regions.png",height = 20, width = 9.6)
```

## INDEL FP and FN calls genome stratification with samples with different purity been merged

```{r, warning=FALSE, message=FALSE, fig.height=7.8, fig.width=9.6}
indel_df |> pivot_longer(c(tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  facet_nested(name ~ cellline+category+tool, nest_line = TRUE,labeller = labeller(name=region_labels, category=category_labels)) + theme_minimal(base_size = 12) +
  labs(x="Proportion", y="Sequencing strategy") + theme(strip.text.y.right = element_text(size=10), axis.text.x.bottom = element_text(size=6))

ggsave("figures/indel_depth_merged_purity_genome_regions.png",height = 7.8,width = 9.6)
```



```{r}
overlap_sv_single <- function(df, grange) {
  df |> mutate(bp1=df[1:3] |> rename(chr1="chr", start1="start",end1="end") |> GRanges() |> countOverlaps(grange),
               bp2=df[4:6] |> rename(chr2="chr", start2="start",end2="end") |> GRanges() |> countOverlaps(grange)) |> 
    mutate(overlap=ifelse(bp1+bp2>0, 1, 0)) |> group_by(supp,overlap) |> summarise(n=n()) |>  ungroup() |> complete(supp,overlap,fill = list(n=0))
}

overlap_sv_with_regions <- function(vcf) {
  tool <- str_extract(vcf, c("delly","nanomonsv","savana","severus") |> paste(collapse = "|"))
  df1 <- read.table(vcf, comment.char = "#") |> dplyr::select(V1,V2) |> 
    mutate(chr1=V1,start1=V2-50, end1=V2+50) |> dplyr::select(chr1,start1,end1)
  df2 <- readVcf(vcf) |> info() |> as.data.frame() |> dplyr::select(CHR2,END,SUPP_VEC) |> 
    mutate(chr2=CHR2, start2=END-50, end2=END+50,supp=SUPP_VEC) |> dplyr::select(chr2,start2,end2,supp)
  df <-  cbind(df1,df2) |> filter(chr1 %in% valid_chr, chr2 %in% valid_chr)
  cds <- overlap_sv_single(df, cds_grange) |> rename(n="refseq_cds")
  lc <- overlap_sv_single(df, lc_grange) |> rename(n="tandem_repeats")
  map <- overlap_sv_single(df, map_grange) |> rename(n="low_mappability")
  gc <- overlap_sv_single(df, gc_grange) |> rename(n="GC")
  segdup <- overlap_sv_single(df, segdup_grange) |> rename(n="segdup")
  full_join(cds,lc, by=c("supp","overlap")) |> full_join(map,by=c("supp","overlap")) |> full_join(gc,by=c("supp","overlap")) |> full_join(segdup,by=c("supp","overlap")) |> add_column(tool=tool)
}
```


```{r,cache=TRUE, message=FALSE,warning=FALSE}
sv_df <- list.files(paste0(workdir,"analysis/benchmark/svs"),  pattern = "*.merged.vcf", full.names = TRUE, recursive = TRUE) |> 
  set_names(\(f) basename(f) |> str_remove(".merged.vcf")) |> map(overlap_sv_with_regions) |> list_rbind(names_to = "prefix") |> 
  separate_wider_delim(prefix,".",names=c("tumor","tumor_depth","normal","normal_depth")) |> 
  separate_wider_delim(tumor, "_",names=c("cellline","purity"), too_few = "align_start") |> 
  mutate(purity=ifelse(is.na(purity),"100",purity), 
         purity=factor(purity, levels=seq(40,100,20)), 
         Overlap=ifelse(overlap==1, "Yes","No"), 
         category=case_when(supp=="01"~"LR novol",supp=="11" ~ "Concordant", supp=="10"~"SR specific"),
         comb=paste(tumor_depth,normal_depth,sep="-")) 
```


## SV FP and FN calls genome stratification

```{r, fig.height=20, fig.width=22}
sv_df  |> 
  pivot_longer(c( tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  facet_nested(name + purity~ cellline+category+tool, nest_line = TRUE,
               labeller = labeller(name=region_labels)) + theme_minimal(base_size = 12) +
  labs(x="Proportion", y="Sequencing strategy") + theme(strip.text.y.right = element_text(size=12), axis.text.x.bottom = element_text(size=6))

ggsave("figures/sv_depth_genome_regions.png",height = 20,width = 22)
```

## SV FP and FN calls genome stratification with samples with different purity been merged

```{r, fig.width=22, fig.height=7.8}
sv_df  |> 
  pivot_longer(c( tandem_repeats,low_mappability,GC, segdup)) |> 
  ggplot(aes(x=value,y=comb,fill=Overlap)) + geom_col(position = "fill") + 
  scale_fill_manual(values = c("#E1BE6A","#40B0A6")) +
  scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1),labels = c(0,0.25,0.5,0.75,1)) +
  facet_nested(name ~ cellline+category+tool, nest_line = TRUE,
               labeller = labeller(name=region_labels)) + theme_minimal(base_size = 12) +
  labs(x="Proportion", y="Sequencing strategy") + theme(strip.text.y.right = element_text(size=10), axis.text.x.bottom = element_text(size=6))

ggsave("figures/sv_depth_merged_purity_genome_regions.png",height = 7.8,width = 22)
```

