---
title: "Check the purity of samples and QC of sequencing data"
output: github_document
---

```{r setup, include=FALSE}
library(here)
library(tidyverse)
library(data.table)
library(ggh4x)
library(patchwork)
#library(VariantAnnotation)
source("config.R")
mypal <- purity_pals

knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center',fig.retina=2)
```

### Sequencing data quality

We sequenced mixed DNA libraries on Promethion, basecalled reads with `dorado`, and aligned them with `minimap2`. The average mapping depth and genomic coverage were computed with `mosdepth` and `bamcov`. The N50 was calcualted with in-house python script [getAlignmentN50.py](scripts/getAlignmentN50.py) using pysam.

```bash
mosdepth -n -t 10 $prefix $input
bamcov -H $input $output
python scripts/getAlignmentN50.py $input > $output
```

```{r, fig.height=6, fig.width=7}
read_bamcov <- function(filename) {
  fread(filename,nrow=23) |> 
  summarise(genome_size=sum(V3), covered_bp = sum(V5), coverage = round(covered_bp/genome_size*100, digits = 3)) |> select(coverage)
}


df_bamcov <- list.files(paste0(workdir,"analysis/qc/bam/R10/sup"), pattern = "*.bamcov.txt", full.names = TRUE) |> 
  set_names(\(x) basename(x) |> str_remove(".bamcov.txt")) |> map(read_bamcov) |> 
  list_rbind(names_to = "sample")

df_mosdepth <- list.files(paste0(workdir,"analysis/qc/bam/R10/sup"), pattern = "*.mosdepth.summary.txt", full.names = TRUE) |>  
  set_names(\(x) basename(x) |> str_remove(".mosdepth.summary.txt")) |> 
  map(\(f) fread(f) |> tail(1) |> select(mean)) |> list_rbind(names_to = "sample")

df_n50 <- list.files(paste0(workdir,"analysis/qc/bam/R10/sup"), pattern = "*.bamN50.txt", full.names = TRUE) |>  
  set_names(\(x) basename(x) |> str_remove(".bamN50.txt")) |> 
  map(\(f) fread(f) |> mutate(n50=V2/1e3) |> select(n50)) |> list_rbind(names_to = "sample") 


named_stats = c("coverage" = "Genome coverage (%)", "mean" = "Mean depth (x)", "n50" = "Alignment N50 (kb)")
df_n50 |> inner_join(df_bamcov) |> inner_join(df_mosdepth) |> separate_wider_delim(sample, delim = "_", names = c("celltype","purity"), too_few ="align_start") |> 
  mutate(purity=ifelse(is.na(purity), 100, purity),
         purity=factor(purity, levels=c(seq(100,10,-10),"BL")),
         sample_type = ifelse(purity=="BL", "B lymphoblast","Tumour")) |> 
  pivot_longer(c(coverage,mean, n50)) |>  
  mutate(name=factor(name,levels=c("mean", "coverage", "n50"))) |> 
  ggplot(aes(x=purity,y=value)) + geom_col(aes(fill=sample_type)) + scale_fill_manual(values = c("#317EC2","#E07F80")) + 
  geom_text(aes(label=round(value, digits = 1)), color='white', size=2.8, vjust=2) +
  facet_grid(name~celltype, scale="free_y",labeller = labeller(name=named_stats)) + theme_bw() + 
  labs(x="Sample purity", y="",fill="") + 
  facetted_pos_scales(
    y = list(
      name == "mean" ~ scale_y_continuous(breaks=c(20,40,60,80), labels = c("20x","40x","60x","80x")),
      name == "coverage" ~ scale_y_continuous(breaks=c(25,50,75,100), labels = c("25%","50%","75%","100%")),
      name == "n50" ~ scale_y_continuous(breaks=c(2.5,5,7.5,10), labels = c("2.5kb","5.0kb","7.5kb","10kb"))
    )
  ) + 
  theme(strip.text = element_text(size=10.5), 
        strip.background = element_rect(fill="white"), 
        axis.text = element_text(size=9),
        legend.position = "bottom")

ggsave("figures/bam_qc.png",height = 6, width = 7)
```
```{r}
df_mosdepth |> separate_wider_delim(sample,"_", names = c("celltype","purity"), too_few = "align_start") |> 
  group_by(celltype) |> summarise(mean_depth=mean(mean), max_depth=max(mean), min_depth=min(mean)) 
```



##Check per flowcell yield for this project

```{r, fig.height=6, fig.width=4}
yield <- read_tsv("../../miscellanous/for_nanopore_catchup_2024April/yield.txt",col_names = c("prefix","gb"))|> 
  separate_wider_delim(prefix,names = c("sample","report"),delim = "/") |> 
  mutate(flowcell=str_split_i(report,"_",i=2)) |> group_by(sample,flowcell) |> mutate(total_gb = sum(gb)) |> 
  ungroup() |> arrange(desc(total_gb)) |> select(-c(gb,report)) |> unique()|> 
  mutate(project=case_when(startsWith(sample,'3300') ~ "Lung",
                           startsWith(sample,'1000') ~ "Lung",
                           startsWith(sample,'COLO829B') ~ "Pilot",
                           sample=="COLO829" ~ "Pilot",
                           sample=="COLO829_BL" ~ "Pilot",
                           TRUE ~ "Mixing"
                   )) |> filter(project=="Mixing",total_gb>0)

n50 <- read_tsv("../../miscellanous/for_nanopore_catchup_2024April/N50.txt", col_names = c("prefix","n50")) |> 
  separate_wider_delim(prefix,names = c("sample","report"),delim = "/") |> 
  mutate(flowcell=str_split_i(report,"_",i=2)) |> group_by(sample,flowcell) |> mutate(mean_len = mean(n50)) |> 
  ungroup() |> dplyr::select(-c(n50,report)) |> unique() |> 
  mutate(project=case_when(startsWith(sample,'3300') ~ "Lung",
                           startsWith(sample,'1000') ~ "Lung",
                           startsWith(sample,'COLO829B') ~ "Pilot",
                           sample=="COLO829" ~ "Pilot",
                           sample=="COLO829_BL" ~ "Pilot",
                           TRUE ~ "Mixing"
                   )) |> filter(project=="Mixing",mean_len>0)

df_summary <- 
  yield |> left_join(n50,by=c("sample","flowcell","project")) |> 
  mutate(type=ifelse(grepl("Normal",sample), "B lymphoblast", "Tumour")) |> 
  arrange(type) |> mutate(flowcell=factor(flowcell,levels=flowcell)) |> 
  pivot_longer(c(total_gb, mean_len)) |> 
  mutate(name=ifelse(name=="mean_len","N50 (kb)","Total yield (gb)"),
         name=factor(name, levels=c("Total yield (gb)","N50 (kb)")))

df_summary_mean <- df_summary |> group_by(name) |> summarise(mean_value=mean(value))

ggplot() + geom_point(data=df_summary, aes(x=value,y=flowcell, color=type)) + 
  geom_vline(data=df_summary_mean,aes(xintercept=mean_value), linetype = "dashed", color="grey") +
  geom_text(data=df_summary_mean, aes(x=mean_value,label=round(mean_value,digits = 2)), y="PAQ29352", hjust=-.1, size=3) +
  facet_wrap(~name, scales = "free_x") +
  theme_minimal() +
  theme_bw() + labs(x="",y="", color="") + scale_color_manual(values = c("#317EC2","#E07F80")) + 
  theme(strip.background = element_rect(fill="white"), 
        strip.text = element_text(size=10),
        axis.text = element_text(size=7), legend.position = "bottom")

ggsave("figures/flowcell_yield.png",height = 6, width = 4)
```



### Check tumor purity 

We used the VAF of all SNPs at CN=2 regions to estimate the purity gradient. The copy-number segments were obtained from short-read data using ascatNGS.

```bash
awk -F"," '{if($7==2) {OFS="\t";print "chr"$2,$3,$4}}' colo829.copynumber.caveman.csv|grep -Ev "chrX|chrY" > colo829.cn2.tsv
awk -F"," '{if($7==2) {OFS="\t";print "chr"$2,$3,$4}}' hcc1937.copynumber.caveman.csv|grep -Ev "chrX|chrY" > hcc1937.cn2.tsv

for i in COLO829*.vcf.gz;do 
    sample=`basename $i .vcf.gz`
    bcftools view -R colo829.cn2.tsv $i > cn2/$sample.vcf
done 

for i in HCC1937*.vcf.gz;do 
    sample=`basename $i .vcf.gz`
    bcftools view -R hcc1937.cn2.tsv $i > cn2/$sample.vcf
done 
```

```{r}
get_vaf <- function(vcf) {
  VariantAnnotation::readVcf(vcf) %>% .[VariantAnnotation::isSNV(.),] |> VariantAnnotation::geno("VAF")|> as.data.frame() |> rownames_to_column("id") |> `colnames<-`(c("id","vaf")) |> mutate(vaf=as.double(vaf))
}

df <- list.files(here("../0.purity_check/dna_mixing/vcf_files/deepsomatic/cn2"), pattern = "vcf$", full.names = TRUE) |> 
  set_names(\(x) basename(x) |> str_remove(".vcf")) |> map(get_vaf) |> list_rbind(names_to = "sample") |>
  mutate(celltype=ifelse(startsWith(sample,"COLO829"),"COLO829","HCC1937"), 
         purity=str_split_i(sample,"_",i=2), 
         purity=ifelse(is.na(purity),"100",purity),
         purity=factor(purity,levels=seq(10,100,10) |> as.character()))
```


```{r fig_1_vaf_curves, fig.height=3.2,fig.width=6.8}
p1 <- df|> filter(celltype=="COLO829") |> 
  ggplot(aes(x=vaf,color=purity)) + 
  stat_density(geom="line",position="identity",linewidth=1) +
  scale_color_manual(values=mypal) + 
  theme_bw(base_size = 12) + 
  labs(x="Allele frequency (AF)",y="Density",color="Purity",title = "COLO829") + 
  theme(legend.position = "none", panel.background = element_rect(fill="lightgrey"))

p2 <- df|> filter(celltype=="HCC1937") |>
  ggplot(aes(x=vaf,color=purity)) + 
  stat_density(geom="line",position="identity",linewidth=1) + 
  scale_color_manual(values=mypal) + 
  theme_bw(base_size = 12) + 
  labs(x="Allele frequency (AF)",y="",color="Purity",title = "HCC1937") + 
  theme(legend.position = "right", panel.background = element_rect(fill="lightgrey"))

p1+p2
ggsave("figures/purity_check_dna_mixing.png", width = 6.8, height = 3.2)
```



