---
title: "Mutational signature analysis of snvs"
output: github_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)
library(deconstructSigs)
#library(VariantAnnotation)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggplot2)
library(ggh4x)
source("config.R")
mypal <- purity_pals #mypal <-scales::seq_gradient_pal("lightblue", "red", "Lab")(seq(0,1,length.out=10))
mypal2 <- colorblind_pals 
names(mypal2) <- c("FP","FN","TP")
chrom_chr <- paste0("chr", c(1:22,"X"))
COLORS6 <- c(
  "#2EBAED", "#000000", "#DE1C14",
  "#D4D2D2", "#ADCC54", "#F0D0CE"
)
knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center', fig.retina=2)
```

Seems HCC1937 had excessive mutations on chromosome X, this is because this cellline has no normal chrX, instead, at least one copy of derived X chromosome. I found most of them could be germline, thus variants on chrX had been excluded in all signature analysis.

```{r}
read_tsv("~/working/general/goldstandard/vcfs/colo829/merged_normed_isec_snv.hom100.goldstandard.vcf.gz", col_names = FALSE, comment = "#") |> 
  dplyr::select(X1, X2, X4, X5) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5) |> add_column(sample="hcc1937") |> filter(chr %in% chrom_chr) |> 
  mutate(chr=factor(chr, levels=chrom_chr)) |> 
  ggplot(aes(x=chr)) + 
  geom_bar()

read_tsv("~/working/general/goldstandard/vcfs/hcc1937/merged_normed_isec_indel.goldstandard.vcf.gz", col_names = FALSE, comment = "#") |> 
  dplyr::select(X1, X2, X4, X5) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5) |> add_column(sample="hcc1937") |> filter(chr %in% chrom_chr) |> 
  mutate(chr=factor(chr, levels=chrom_chr)) |> 
  ggplot(aes(x=chr)) + 
  geom_bar()
```

```{r}
clairS_lr <- list.files("~/working/bioprojects/nanopore_celllines_benchmark/1.dna_mixing_celllines/work/analysis/snvs/clairS/R10/sup", pattern = "output.vcf.gz$", full.names = TRUE, recursive = TRUE) |> 
  set_names(\(f) dirname(f)) |> 
  map(\(f) read_tsv(f, col_names=FALSE,comment = "#") |> dplyr::select(X1, X2, X4, X5, X6) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5, "qual"=X6)  ) |> 
  list_rbind(names_to = "prefix") |> 
  mutate(sample=basename(prefix) |> str_split_i("\\.",i=1), tool=str_extract(prefix,"deepsomatic|clairS"), purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity))|> dplyr::select(-prefix)

dpsomatic_lr <- list.files("~/working/bioprojects/nanopore_celllines_benchmark/1.dna_mixing_celllines/work/analysis/snvs/deepsomatic//R10/sup", pattern = "output.somatic.vcf.gz$", full.names = TRUE, recursive = TRUE) |> 
  set_names(\(f) dirname(f)) |> 
  map(\(f) read_tsv(f, col_names=FALSE,comment = "#") |> dplyr::select(X1, X2, X4, X5, X6) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5, "qual"=X6)  ) |> 
  list_rbind(names_to = "prefix") |> 
  mutate(sample=basename(prefix) |> str_split_i("\\.",i=1), tool=str_extract(prefix,"deepsomatic|clairS"), purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity))|> dplyr::select(-prefix)
    
clairS_lr_mut <- clairS_lr |> mutate(celltype=ifelse(startsWith(sample,"COLO829"), "COLO829", "HCC1937")) |> 
  filter(chr %in% chrom_chr) |>
  filter(!(celltype=="HCC1937" & chr=="chrX")) |>  # exclude mutations on chrX in HCC1937
  unite("sample_id", c(sample,purity,tool), sep = "@") |>
  dplyr::select(sample_id, chr, pos, ref, alt)|> as.data.frame() |> mut.to.sigs.input(sample.id = "sample_id", bsg=BSgenome.Hsapiens.UCSC.hg38)

dpsomatic_lr_mut <- dpsomatic_lr |>  mutate(celltype=ifelse(startsWith(sample,"COLO829"), "COLO829", "HCC1937")) |> 
  filter(chr %in% chrom_chr) |> 
  filter(!(celltype=="HCC1937" & chr=="chrX")) |>
  unite("sample_id", c(sample,purity,tool), sep = "@") |>
  dplyr::select(sample_id, chr, pos, ref, alt)|> as.data.frame() |> mut.to.sigs.input(sample.id = "sample_id", bsg=BSgenome.Hsapiens.UCSC.hg38)
```


### The mutational spectrum in COLO829 gold standard

```{r, fig.height=2.6, fig.width=10.6}
gc_colo829 <- read_tsv("~/working/general/goldstandard/vcfs/colo829/merged_normed_isec_snv.hom100.goldstandard.vcf.gz", col_names = FALSE, comment = "#") |> 
dplyr::select(X1, X2, X4, X5) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5) |> add_column(sample="colo829") |> 
  as.data.frame() |> mut.to.sigs.input(sample.id = "sample", bsg=BSgenome.Hsapiens.UCSC.hg38)

gc_colo829 |> rownames_to_column("sample") |> pivot_longer(-sample) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> 
  mutate(prop=value/sum(value)) |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")


ggsave("figures/mut_cat_colo829_gc.png", width = 10.6, height = 2.6)
```


### The mutational spectrum in long read calls for COLO829

```{r, fig.height=10.6, fig.width=10.6}
clairS_lr_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"COLO829")) |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context", y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_colo829_clairS.png", width = 10.6, height = 8.6)

dpsomatic_lr_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"COLO829")) |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context", y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_colo829_dpsomatic.png", width = 10.6, height = 8.6)
```

The signatures in samples of different purity levels

```{r}
cosmicv3_h38  <- read_table("data/COSMIC_v3.4_SBS_GRCh38.txt") |> column_to_rownames("Type")
cosmicv3_h38 <- cosmicv3_h38[match(colnames(dpsomatic_lr_mut), rownames(cosmicv3_h38)),] |> t() |> as.data.frame()
cosmicv2_h38 <- read_table("../../../breast_cancer_tms_Amanda2023/brca_mutational_signatures/data/COSMIC/COSMIC_v2_SBS_GRCh38.txt") |> column_to_rownames("Type")
cosmicv2_h38 <- cosmicv2_h38[match(colnames(dpsomatic_lr_mut), rownames(cosmicv2_h38)),] |> t() |> as.data.frame()

#clairS_sigs <- rownames(clairS_lr_mut) |> 
#  map(\(x)
#    whichSignatures(tumor.ref = clairS_lr_mut, 
#                       signatures.ref = cosmicv3_h38, 
#                       sample.id = x,
#                       contexts.needed = TRUE) %>% .$weights
#  ) |> list_rbind()
#
#dpsomatic_sigs <- rownames(dpsomatic_lr_mut) |> 
#  map(\(x)
#    whichSignatures(tumor.ref = dpsomatic_lr_mut, 
#                       signatures.ref = cosmicv3_h38, 
#                       sample.id = x,
#                       contexts.needed = TRUE) %>% .$weights
#  ) |> list_rbind()

clairS_sigs_v2 <- rownames(clairS_lr_mut) |> 
  map(\(x)
    whichSignatures(tumor.ref = clairS_lr_mut, 
                       signatures.ref = cosmicv2_h38, 
                       sample.id = x,
                       contexts.needed = TRUE) %>% .$weights
  ) |> list_rbind()

dpsomatic_sigs_v2 <- rownames(dpsomatic_lr_mut) |> 
  map(\(x)
    whichSignatures(tumor.ref = dpsomatic_lr_mut, 
                       signatures.ref = cosmicv2_h38, 
                       sample.id = x,
                       contexts.needed = TRUE) %>% .$weights
  ) |> list_rbind()


sigs_v2_included <- unique(clairS_sigs_v2 |> colSums() %>% {which(. >= .05)} |> names(), dpsomatic_sigs_v2 |> colSums() %>% {which(. >= .05)} |> names())
sig_v2_pal <- RColorBrewer::brewer.pal(11,"Set3")
names(sig_v2_pal) <- sigs_v2_included
```

```{r}
clairS_sigs_v2 |> rownames_to_column("sample_id") |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  filter(startsWith(sample,"COLO829")) |> 
  pivot_longer(-c(sample, purity, tool)) |> 
  filter(value!=0) |> 
  ggplot(aes(x=purity,y=value,fill=name)) + 
  geom_col(position = "fill", color="black", linewidth=.2) + 
  scale_fill_manual(values=sig_v2_pal) +
  theme_classic() + labs(x="Purity", y="Proportion", fill="COSMIC signatures") 

ggsave("figures/colo829_clairS_sigs_v2.png",height = 3,width = 8)

dpsomatic_sigs_v2 |> rownames_to_column("sample_id") |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  filter(startsWith(sample,"COLO829")) |> 
  pivot_longer(-c(sample, purity, tool)) |> 
  filter(value!=0) |> 
  ggplot(aes(x=purity,y=value,fill=name)) + 
  geom_col(position = "fill", color="black", linewidth=.2) + 
  scale_fill_manual(values=sig_v2_pal) +
  theme_classic() + labs(x="Purity", y="Proportion", fill="COSMIC signatures")

ggsave("figures/colo829_dpsomatic_sigs_v2.png",height = 3, width = 8)
```

### The mutational spectrum in false positive calls

```{r, fig.height=10.6, fig.width=10.6}
read_bc_vcf <- function(filename) {
  sample <- dirname(filename) |> basename() |> str_split_i("\\.",i=1)
  tool <- filename |> str_extract("deepsomatic|clairS")
  if(tool=="deepsomatic") {
    info_cols <- c("GT","GQ","DP","AD","VAF","PL" )
    af_col <- "VAF"
  } else {
    info_cols <- c("GT","GQ","DP","AF","AD","NAF","NDP","NAD","AU","CU","GU","TU","NAU","NCU","NGU","NTU")
    af_col <- "AF"
  }
  
  read_tsv(filename, col_names = F) |> separate_wider_delim(cols = "X10",delim = ":",names = info_cols) |>
    add_column(sample=sample,tool=tool) |> 
    dplyr::select(X1, X2, X4, X5, X6, GT, DP, af_col, sample, tool) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5,"qual" = X6,"vaf"=af_col) |> 
    mutate(purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity),
           purity=factor(purity,levels=seq(10,100,10) |> as.character()))
}

df_fp <- list.files(paste0(workdir, "analysis/benchmark/snvs/somatic/R10/sup"), pattern = "fp.vcf", recursive = TRUE,full.names = T) |>
  keep(\(x) grepl("passed",x)) |> discard(\(x) grepl("noXY",x)) |> 
  set_names(\(f) dirname(f) |> basename() |> str_split_i("\\.",i=1)) |> 
  map(read_bc_vcf) |> list_rbind()  |> add_column(type="FP")

fp_mut <- df_fp |> mutate(celltype=ifelse(startsWith(sample,"COLO829"), "COLO829", "HCC1937")) |> 
  filter(chr %in% chrom_chr) |>
  filter(!(celltype=="HCC1937" & chr=="chrX")) |>  # exclude mutations on chrX in HCC1937
  unite("sample_id", c(sample,purity,tool), sep = "@") |>
  dplyr::select(sample_id, chr, pos, ref, alt)|> as.data.frame() |> mut.to.sigs.input(sample.id = "sample_id", bsg=BSgenome.Hsapiens.UCSC.hg38)

fp_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"COLO829"),tool=="clairS") |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_colo829_clairS_fp.png", width = 10.6, height = 8.6)

fp_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"COLO829"),tool=="deepsomatic") |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_colo829_dpsomatic_fp.png", width = 10.6, height = 8.6)
```

### The mutational spectrum in HCC1937 gold standard

```{r,fig.height=2.6, fig.width=10.6}
gc_hcc1937 <- read_tsv("~/working/general/goldstandard/vcfs/hcc1937/merged_normed_isec_snv.hom100.goldstandard.vcf.gz", col_names = FALSE, comment = "#") |> 
dplyr::select(X1, X2, X4, X5) |> dplyr::rename("chr"=X1,"pos"=X2, "ref"=X4, "alt"=X5) |> add_column(sample="hcc1937") |> filter(chr%in% chrom_chr, chr!="chrX") |> 
  as.data.frame() |> mut.to.sigs.input(sample.id = "sample", bsg=BSgenome.Hsapiens.UCSC.hg38)

gc_hcc1937 |> rownames_to_column("sample") |> pivot_longer(-sample) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> 
  mutate(prop=value/sum(value)) |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_hcc1937_gc.png", width = 10.6, height = 2.6)
```


### The mutational spectrum in long read calls

```{r, fig.height=10.6, fig.width=10.6}
clairS_lr_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"HCC1937")) |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_hcc1937_clairS.png", width = 10.6, height = 8.6)

dpsomatic_lr_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"HCC1937")) |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_hcc1937_dpsomatic.png", width = 10.6, height = 8.6)
```

```{r}
clairS_sigs_v2 |> rownames_to_column("sample_id") |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  filter(startsWith(sample,"HCC1937")) |> 
  pivot_longer(-c(sample, purity, tool)) |> 
  filter(value!=0) |> 
  ggplot(aes(x=purity,y=value,fill=name)) +
  geom_col(position = "fill", color="black", linewidth=.2) + 
  scale_fill_manual(values=sig_v2_pal) +
  theme_classic() + labs(x="Purity", y="Proportion", fill="COSMIC signatures")

ggsave("figures/hcc1937_clairS_sigs_v2.png",height = 3,width = 8)

dpsomatic_sigs_v2 |> rownames_to_column("sample_id") |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  filter(startsWith(sample,"HCC1937")) |> 
  pivot_longer(-c(sample, purity, tool)) |> 
  filter(value!=0) |> 
  ggplot(aes(x=purity,y=value,fill=name)) + 
  geom_col(position = "fill", color="black", linewidth=.2) + 
  scale_fill_manual(values=sig_v2_pal) +
  theme_classic() + labs(x="Purity", y="Proportion", fill="COSMIC signatures")

ggsave("figures/hcc1937_dpsomatic_sigs_v2.png",height = 3,width = 8)
```

### The mutational spectrum in false positive calls in HCC1937

```{r, fig.height=10.6, fig.width=10.6}
fp_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"HCC1937"),tool=="clairS") |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_hcc1937_clairS_fp.png", width = 10.6, height = 8.6)

fp_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> filter(startsWith(sample,"HCC1937"),tool=="deepsomatic") |> 
  group_by(purity) |> mutate(prop=value/sum(value)) |> ungroup() |> 
  ggplot(aes(x=context,y=prop,fill=mut)) + geom_col(color="black",linewidth=.1) +
  scale_fill_manual(values = COLORS6) + 
  facet_grid(purity~mut, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90, size=7), strip.background.y = element_blank()) +
  labs(x="Context",y="Number of single base substitution",fill="Mutation type")

ggsave("figures/mut_cat_hcc1937_dpsomatic_fp.png", width = 10.6, height = 8.6)
```


### HRD prediction using samples with different purity

```{r}
library(CHORD)
library(VariantAnnotation)

chord_hrd_clairS_df <- function(sample) {
  snv_df <- readVcf(paste0("../work/analysis/snvs/clairS/R10/sup/",sample,".HCC1937_BL/output.no_chrXY.vcf.gz"), genome = "hg38") %>%   { .[fixed(.)$FILTER == "PASS", ] } %>% 
  { .[isSNV(.), ] } %>%  
  {
    data.frame(
      chrom = seqnames(rowRanges(.)) |> as.character(),  
      pos = start(rowRanges(.)),       
      ref = as.character(ref(.)),      
      alt = sapply(alt(.), as.character)
    )
  }

indel_df <- readVcf(paste0("../work/analysis/snvs/clairS/R10/sup/",sample,".HCC1937_BL/norm_indel.no_chrXY.vcf.gz"), genome = "hg38") %>% { .[fixed(.)$FILTER == "PASS", ] } %>%
  { .[!isSNV(.), ] } %>%  
  {
    data.frame(
      chrom = seqnames(rowRanges(.)) |> as.character(),  
      pos = start(rowRanges(.)),       
      ref = as.character(ref(.)),      
      alt = sapply(alt(.), as.character)    
    )
  }

sv_df <- readVcf(paste0("../work/analysis/svs/simple/severus/R10/sup/",sample,".HCC1937_BL.vcf"), genome = "hg38") %>% 
  { .[seqnames(rowRanges(.)) != "chrX", ] } |> info() |> as.data.frame() |> dplyr::select(SVTYPE, SVLEN) |> 
  rename(SVTYPE="sv_type", SVLEN="sv_len") |> mutate(sv_len=ifelse(sv_type=="TRA",NA, as.numeric(sv_len)))

extractSigsChord(df.snv = snv_df,
                 df.indel = indel_df,
                 df.sv = sv_df,
                 sample.name = sample,
                 ref.genome = BSgenome.Hsapiens.UCSC.hg38) |> 
  chordPredict(do.bootstrap = T, verbose = F)
}


hcc1937_clairS_chord_hrd <- paste0("HCC1937",c("",paste0("_",seq(10,90,10)))) |> map(\(x) chord_hrd_clairS_df(x)) |> list_rbind()


chord_hrd_dpsomatic_df <- function(sample) {
  snv_df <- readVcf(paste0("../work/analysis/snvs/deepsomatic/R10/sup/",sample,".HCC1937_BL/output.somatic.no_chrXY.vcf.gz"), genome = "hg38") %>%   { .[fixed(.)$FILTER == "PASS", ] } %>% 
  { .[isSNV(.), ] } %>%  
  {
    data.frame(
      chrom = seqnames(rowRanges(.)) |> as.character(),  
      pos = start(rowRanges(.)),       
      ref = as.character(ref(.)),      
      alt = sapply(alt(.), as.character)
    )
  }

indel_df <- readVcf(paste0("../work/analysis/snvs/clairS/R10/sup/",sample,".HCC1937_BL/norm_indel.no_chrXY.vcf.gz"), genome = "hg38") %>% { .[fixed(.)$FILTER == "PASS", ] } %>%
  { .[!isSNV(.), ] } %>%  
  {
    data.frame(
      chrom = seqnames(rowRanges(.)) |> as.character(),  
      pos = start(rowRanges(.)),       
      ref = as.character(ref(.)),      
      alt = sapply(alt(.), as.character)    
    )
  }

sv_df <- readVcf(paste0("../work/analysis/svs/simple/severus/R10/sup/",sample,".HCC1937_BL.vcf"), genome = "hg38") %>% 
  { .[seqnames(rowRanges(.)) != "chrX", ] } |> info() |> as.data.frame() |> dplyr::select(SVTYPE, SVLEN) |> 
  rename(SVTYPE="sv_type", SVLEN="sv_len") |> mutate(sv_len=ifelse(sv_type=="TRA",NA, as.numeric(sv_len)))

extractSigsChord(df.snv = snv_df,
                 df.indel = indel_df,
                 df.sv = sv_df,
                 sample.name = sample,
                 ref.genome = BSgenome.Hsapiens.UCSC.hg38) |> 
  chordPredict(do.bootstrap = T, verbose = F)
}

hcc1937_dpsomatic_chord_hrd <- paste0("HCC1937",c("",paste0("_",seq(10,90,10)))) |> map(\(x) chord_hrd_dpsomatic_df(x)) |> list_rbind()
```

```{r}
hcc1937_clairS_chord_hrd |> dplyr::select(sample, p_BRCA1, p_BRCA2) |> 
  mutate(purity=str_split_i(sample,"_",i=2),purity=ifelse(is.na(purity),"100",purity),purity=factor(purity, levels=seq(100,10,-10))) |> 
  pivot_longer(c(p_BRCA1,p_BRCA2)) |> mutate(name=factor(name, levels=c("p_BRCA1","p_BRCA2"))) |> 
  ggplot(aes(x=purity, y=value, fill=name)) + 
  geom_col(aes(order=fct_rev(name)),position = "stack") + theme_bw() + ggsci::scale_fill_jama(labels=c("BRCA1","BRCA2")) +
  geom_hline(yintercept = 0.5, color="darkred", linetype=2, linewidth=1.5) +
  labs(x="Purity", y="Probability", fill="BRCA1/2 type")

ggsave("figures/hcc1937_clairS_chord_hrd.png",height = 3,width = 8)

hcc1937_dpsomatic_chord_hrd |> dplyr::select(sample, p_BRCA1, p_BRCA2) |> 
  mutate(purity=str_split_i(sample,"_",i=2),purity=ifelse(is.na(purity),"100",purity),purity=factor(purity, levels=seq(100,10,-10))) |> 
  pivot_longer(c(p_BRCA1,p_BRCA2)) |> mutate(name=factor(name, levels=c("p_BRCA1","p_BRCA2"))) |> 
  ggplot(aes(x=purity, y=value, fill=name)) + 
  geom_col(aes(order=fct_rev(name)),position = "stack") + theme_bw() + ggsci::scale_fill_jama(labels=c("BRCA1","BRCA2")) +
  geom_hline(yintercept = 0.5, color="darkred", linetype=2, linewidth=1.5) +
  labs(x="Purity", y="Probability", fill="BRCA1/2 type")

ggsave("figures/hcc1937_dpsomatic_chord_hrd.png",height = 3,width = 8)
```

```{r}
chord_hrd_fp <- function(sample, tool) {
  snv_df <- read_tsv(paste0("../work/analysis/benchmark/snvs/somatic/R10/sup/",tool,"/",sample,".HCC1937_BL_passed/fp.vcf"), col_names = c("chrom","pos","id","ref","alt")) |> 
  dplyr::select(chrom,pos, ref,alt) |> filter(chrom != "chrX")

  indel_df <- read_tsv(paste0("../work/analysis/benchmark/indels/somatic/R10/sup/",tool,"/", sample, ".HCC1937_BL/fp.vcf"), col_names = c("chrom","pos","id","ref","alt")) |> 
  dplyr::select(chrom,pos, ref,alt) |> filter(chrom != "chrX")

  sv_df <- readVcf(paste0("../work/analysis/benchmark/svs/severus/R10/sup/", sample, ".HCC1937_BL.merged.vcf"), genome = "hg38") %>% 
  { .[seqnames(rowRanges(.)) != "chrX", ] } %>%
  { .[info(.)$SUPP_VEC == "01", ] } |> 
  info() |> as.data.frame() |> dplyr::select(SVTYPE, SVLEN) |> 
  rename(SVTYPE="sv_type", SVLEN="sv_len") |> mutate(sv_len=ifelse(sv_type=="TRA",NA, as.numeric(sv_len)))
  
  extractSigsChord(df.snv = snv_df,
                 df.indel = indel_df,
                 df.sv = sv_df,
                 sample.name = sample,
                 ref.genome = BSgenome.Hsapiens.UCSC.hg38) |> 
  chordPredict(do.bootstrap = T, verbose = F)
}

hcc1937_clairS_fp_chord_hrd <- paste0("HCC1937",c("",paste0("_",seq(10,90,10)))) |> map(\(x) chord_hrd_fp(x, tool="clairS")) |> list_rbind()
hcc1937_dpsomatic_fp_chord_hrd <- paste0("HCC1937",c("",paste0("_",seq(10,90,10)))) |> map(\(x) chord_hrd_fp(x, tool="deepsomatic")) |> list_rbind()


hcc1937_clairS_fp_chord_hrd |> dplyr::select(sample, p_BRCA1, p_BRCA2) |> 
  mutate(purity=str_split_i(sample,"_",i=2),purity=ifelse(is.na(purity),"100",purity),purity=factor(purity, levels=seq(100,10,-10))) |> 
  pivot_longer(c(p_BRCA1,p_BRCA2)) |> mutate(name=factor(name, levels=c("p_BRCA1","p_BRCA2"))) |> 
  ggplot(aes(x=purity, y=value, fill=name)) + 
  geom_col(aes(order=fct_rev(name)),position = "stack") + theme_bw() + ggsci::scale_fill_jama(labels=c("BRCA1","BRCA2")) +
  geom_hline(yintercept = 0.5, color="darkred", linetype=2, linewidth=1.5) +
  labs(x="Purity", y="Probability", fill="BRCA1/2 type")

ggsave("figures/hcc1937_clairS_fp_chord_hrd.png",height = 3,width = 8)

hcc1937_dpsomatic_fp_chord_hrd |> dplyr::select(sample, p_BRCA1, p_BRCA2) |> 
  mutate(purity=str_split_i(sample,"_",i=2),purity=ifelse(is.na(purity),"100",purity),purity=factor(purity, levels=seq(100,10,-10))) |> 
  pivot_longer(c(p_BRCA1,p_BRCA2)) |> mutate(name=factor(name, levels=c("p_BRCA1","p_BRCA2"))) |> 
  ggplot(aes(x=purity, y=value, fill=name)) + 
  geom_col(aes(order=fct_rev(name)),position = "stack") + theme_bw() + ggsci::scale_fill_jama(labels=c("BRCA1","BRCA2")) +
  geom_hline(yintercept = 0.5, color="darkred", linetype=2, linewidth=1.5) +
  labs(x="Purity", y="Probability", fill="BRCA1/2 type")

ggsave("figures/hcc1937_dpsomatic_fp_chord_hrd.png",height = 3,width = 8)
```


### Extra: The signature analysis with FP SNVs

This result might be not that useful, since FP in callset will cause overfitting. So should not included this is paper.

```{r}
fp_sigs_v2 <- rownames(fp_mut) |> 
  map(\(x)
    whichSignatures(tumor.ref = fp_mut, 
                       signatures.ref = cosmicv2_h38, 
                       sample.id = x,
                       contexts.needed = TRUE) %>% .$weights
  ) |> list_rbind()

extra_fp_sigs <- setdiff(fp_sigs_v2 |> colSums() %>% {which(. >.05)} |> names(), sigs_v2_included)

fp_sig_v2_pal <- c(sig_v2_pal, RColorBrewer::brewer.pal(9,"Set1"))
names(fp_sig_v2_pal) <- c(sigs_v2_included, extra_fp_sigs)
```

```{r}
fp_sigs_v2 |> rownames_to_column("sample_id") |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(celltype=ifelse(startsWith(sample,"COLO829"),"COLO829","HCC1937"),purity=factor(purity, levels=seq(100,10,-10))) |> 
  #filter(startsWith(sample,"COLO829")) |> 
  pivot_longer(-c(sample, purity, tool, celltype)) |> 
  filter(value!=0) |> 
  ggplot(aes(x=purity,y=value,fill=name)) + 
  geom_col(position = "fill", color="black", linewidth=.2) + 
  scale_fill_manual(values=fp_sig_v2_pal) +
  theme_bw() + labs(x="Purity", y="Proportion", fill="COSMIC signatures") +
  facet_grid(celltype~tool, scales = "free")
```


```{r, fig.height=6.7, fig.width=5.8}
custom_x <- list(
  scale_x_continuous(limits = c(0, 0.22)),
  scale_x_continuous(limits = c(0, 0.07)),
  scale_x_continuous(limits = c(0, 0.22)),
  scale_x_continuous(limits = c(0, 0.08))
)

custom_y <- list(
  scale_y_continuous(limits = c(0, 0.22)),
  scale_y_continuous(limits = c(0, 0.07)),
  scale_y_continuous(limits = c(0, 0.22)),
  scale_y_continuous(limits = c(0, 0.08))
)

gc_mut <- rbind(gc_colo829, gc_hcc1937) |> rownames_to_column("celltype") |> mutate(celltype=toupper(celltype))|> pivot_longer(-celltype) |> group_by(celltype) |> mutate(gc_prop=value/sum(value)) |> ungroup() |> dplyr::select(-value)

fp_mut_diff <- fp_mut |> rownames_to_column("sample_id") |> pivot_longer(-sample_id) |> 
  separate_wider_delim(sample_id, names= c("sample","purity","tool"), delim = "@") |> 
  mutate(purity=factor(purity, levels=seq(100,10,-10))) |> 
  mutate(c1=str_split_i(name, "\\[", i=1), c2=str_split_i(name,"\\]",i=2), mut=sub(".*\\[(.*?)\\].*", "\\1", name)) |> 
  unite("context",c(c1,c2), sep = "") |> group_by(sample,tool) |> mutate(sum=sum(value),prop=value/sum) |> ungroup() |> 
  mutate(celltype=ifelse(startsWith(sample,"COLO829"), "COLO829","HCC1937")) |> 
  full_join(gc_mut, by=c("name","celltype")) |> mutate(rel_diff=(prop-gc_prop)/gc_prop, diagonal_dist=abs(prop-gc_prop)/sqrt(2)) |> 
  filter(purity=="100")

fp_mut_diff |> ggplot(aes(x=gc_prop, y=prop)) + geom_point(aes(color=mut), alpha=.6, size=2.5) + 
    ggrepel::geom_text_repel(data=fp_mut_diff|> filter(rel_diff>0) |> group_by(celltype, tool) |> slice_max(diagonal_dist, n=3), aes(label=context, color=mut), show.legend = FALSE, fontface="bold") +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "darkgrey") + 
    geom_label(data=fp_mut_diff |> group_by(purity, sample, celltype, tool) |> summarise(cosine_sim=sum(prop*gc_prop)/(sqrt(sum(prop*prop)) * sqrt(sum(gc_prop*gc_prop)))) |> ungroup(),
               aes(label=paste0("cosine sim=",round(cosine_sim,digits = 2)), x = -Inf, y = Inf), vjust=1.5, hjust=-.08, color="darkblue") + 
    #ggpubr::stat_cor(aes(label = after_stat(rr.label)), color = "darkblue", geom = "label", method = "spearman") + 
    facet_wrap(tool~celltype, scales = "free", labeller = labeller(tool=public_tool_names)) + scale_color_manual(values=COLORS6) + theme_classic() +
    facetted_pos_scales(x = custom_x, y=custom_y) + theme(aspect.ratio = 1) + theme_bw() + theme(strip.background =element_rect(fill="white"), legend.position = "bottom") +
    labs(x="Mutational catalogue in gold standard", y="Mutational catalogue in FP", color="")

ggsave("figures/figure5a.pdf",height = 6.7, width = 5.8, dpi = "retina")
```
