---
title: "SV calling"
output: github_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)
library(gtools)
library(UpSetR)
library(ggh4x)
library(VariantAnnotation)
source("config.R")
mypal <- colorblind_pals
names(mypal) <- c("01","10","11")
#sv_palette = c( "DEL"="#BC4749","INS/DUP"="#F4ACB7","INV"="#588157","TRA" = "#FF9F1C")
sv_palette = c( "DEL"="#0081a7","INS/DUP"="#F4ACB7","INV"="#588157","TRA" = "#FF9F1C")

knitr::opts_chunk$set(root.dir = here(), comment = NA, echo = FALSE, message = FALSE, warnings = FALSE, include = TRUE, fig.align = 'center',fig.retina=2)
```

We benchmarked the SV calling over purities using 4 somatic SV callers: nanomonsv, severus, savana, and delly.

**1.delly**

```bash
export OMP_NUM_THREADS=2 # delly primarily parallises on the sample level.
delly lr \
            -t ALL \
            -y ont \
            -o $output_bcf \
            -g $genome $tumor_bam $normal_bam
        
function get_sample_id() {
      echo "$(samtools view -H ${1} | perl -lne 'print ${1} if /\\sSM:(\\S+)/' | head -n 1 )"
}

TID=$(get_sample_id "$tumour_bam")
CID=$(get_sample_id "$normal_bam")
printf "${TID}\\ttumor\\n${CID}\\tcontrol\\n" > $samples_tsv

delly filter \
      -f somatic \
      -p \
      -s $samples_tsv \
      $output_bcf | bcftools view -Ov > $output_vcf
```

**2.nanomonsv**

```bash
nanomonsv parse $tumor_bam $tumor
nanomonsv parse $normal_bam $normal

nanomonsv get $tumor $tumor_bam $genome \
        --control_prefix $normal --control_bam $normal_bam \
        --single_bnd --use_racon --min_indel_size 10 --qv20 \
        --control_panel_prefix $control_panel --processes 24
        
python3 misc/add_simple_repeat.py $input $ouptut $simple_repeat

head -n 1 $output > $passed_output 
tail -n +2 $output |grep PASS >> $passed_output

python3 misc/sv_type.py $passed_output $passed_outptu_with_svtype
```

**3.SAVANA***

```bash
savana --tumour $tumour_bam \
        --normal $normal_bam \
        --ref $genome \
        --sample $tumor.$normal \
        --threads 24 \
        --outdir $outdir
```

**4.Severus**

```bash
 severus --target-bam $hp_tagged_tumour_bam \
            --control-bam $hp_tagged_normal_bam \
            --out-dir $outdir \
            --threads 24 \
            --vntr-bed $vntr_bed \
            --phasing-vcf $phased_vcf
```

```{r}
read_merged_vcf <- function(filename) {
  tool <- str_extract(filename, c("delly","severus","savana","nanomonsv") |> paste(collapse = "|"))
  sample <- basename(filename) |> str_remove(".merged.vcf") |> str_split_i("\\.",i=1)
  data.table::fread(cmd = paste0("grep -v '##' ", filename, " | grep -E '^#|SUPP_VEC=10|SUPP_VEC=11' "), 
        header = TRUE,sep = "\t") |> 
    mutate(supp=str_split_i(INFO,";",i=18) |> str_remove("SUPP="), 
           svtype=str_split_i(INFO, ";", i=1) |> str_remove("SVTYPE=")) |> 
    dplyr::select(c(`#CHROM`, POS, ID, svtype,supp)) |> add_column(sample=sample, tool=tool)
}

df_sv <- list.files(paste0(workdir, "analysis/benchmark/svs"), pattern = "*\\.merged.vcf",recursive = TRUE, full.names = TRUE) |> map(read_merged_vcf) |> list_rbind() |>  mutate(purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity), cellline=ifelse(startsWith(sample,'COLO829'),"COLO829","HCC1937"), svtype=ifelse(svtype %in% c("INS","DUP"), "INS/DUP",svtype)) 

id_levels <- df_sv |> arrange(svtype, factor(`#CHROM`, levels=mixedsort(`#CHROM`) |> unique())) |> pull(ID) |> unique()
```

### COLO829

```{r, fig.height=5.6, fig.width=8}
p1 <- df_sv |> filter(cellline=="COLO829")|>  mutate(ID=factor(ID,levels=id_levels), purity=factor(purity,levels=seq(10,100,10) |> as.character())) |> 
  ggplot(aes(x=ID,y=purity,fill=svtype, alpha=supp)) + 
  geom_tile(color="black") + 
  facet_grid2(tool~svtype, scales = "free", space = "free",labeller = labeller(tool=public_tool_names),
              strip = strip_themed(background_x = elem_list_rect(fill=sv_palette),
                                   background_y = elem_list_rect(fill="white"))
              ) +
  scale_alpha_manual(values = c(0.1, 1), labels = c("Missed", "Detected")) + 
  scale_fill_manual(values = sv_palette) + theme_bw(base_size = 12) +
  theme(axis.text.x = element_text(angle=45, hjust=1),
    legend.position="bottom",
    legend.direction = "horizontal",     
    legend.margin = margin(0, 0, 0, 0), 
    axis.text.x.bottom = element_blank(), 
    axis.ticks.x = element_blank(),
    strip.text = element_text(size=10), 
    axis.text = element_text(size=8))+  # Remove margins around the legend
  #scale_x_discrete(labels = function(x) str_remove(x, "0_0_truthset_")) +
  labs(x="SV gold standard", y="Tumour Purity", alpha="", fill="SV type")

p1
p_legend <- cowplot::get_legend(p1)
ggsave("figures/sv_legend.png", p_legend)
p1 <- p1 + theme(legend.position = "none")
ggsave("figures/sv_heatmap_colo829.png", height = 5.6, width = 8)
```


### HCC1937

```{r, fig.height=5.6, fig.width=12}
p2 <- df_sv|> filter(cellline=="HCC1937")|>  mutate(ID=factor(ID,levels=id_levels), purity=factor(purity,levels=seq(10,100,10) |> as.character())) |> 
  ggplot(aes(x=ID,y=purity,fill=svtype, alpha=supp)) + 
  geom_tile(color="black") + 
  facet_grid2(tool~svtype, scales = "free", space = "free",labeller = labeller(tool=public_tool_names),
              strip = strip_themed(background_x = elem_list_rect(fill=sv_palette),
                                   background_y = elem_list_rect(fill="white"))
              ) +
  scale_alpha_manual(values = c(0.1, 1), labels = c("Missed", "Detected")) + 
  scale_fill_manual(values = sv_palette) + theme_bw() +
  theme(axis.text.x = element_text(angle=45,hjust=1),
    legend.position="bottom",
    legend.direction = "horizontal",     
    legend.margin = margin(0, 0, 0, 0), 
    axis.text.x.bottom = element_blank(),  
    axis.ticks.x = element_blank(),
    strip.text = element_text(size=10), 
    axis.text = element_text(size=8))+ 
  labs(x="SV gold standard", y="Tumour Purity", alpha="")

p2
p2 <- p2 + theme(legend.position = "none")
ggsave("figures/sv_heatmap_hcc1937.png", height = 5.6,width = 16)
```


Try facet_nested

```{r, fig.height=5.6, fig.width=16}
panelsize_count <- df_sv |> group_by(cellline, svtype, sample, tool) |> summarise(n=n()) |> ungroup() |> dplyr::select(cellline, svtype,n)  |> unique() |> pull(n)
colo829_panel <- panelsize_count[1:4]/min(panelsize_count[1:4])
hcc1937_panel <- panelsize_count[5:8]/min(panelsize_count[5:8]) * 3 ## hand-made panel size in proportion. but I need hcc1937 to be slightly longer but not proportional to the total number.

df_sv|>  mutate(ID=factor(ID,levels=id_levels), purity=factor(purity,levels=seq(10,100,10) |> as.character())) |> 
  ggplot(aes(x=ID,y=purity,fill=svtype, alpha=supp)) + 
  geom_tile(color="black") + 
  facet_nested(tool~cellline + svtype, scale="free_x", space = "free_x", 
               strip = strip_nested(background_x = elem_list_rect(fill=c(rep("white",2), rep(sv_palette,2))), 
                                    background_y = elem_list_rect(fill="white")),labeller = labeller(tool=public_tool_names)) +
  force_panelsizes(cols = c(colo829_panel, hcc1937_panel),rows = c(2,2,2,2), respect = TRUE) + 
  scale_alpha_manual(values = c(0.1, 1), labels = c("Missed", "Detected")) + 
  scale_fill_manual(values = sv_palette) + theme_bw() +
  theme(axis.text.x = element_text(angle=45,hjust=1),
    legend.position="none",
    legend.margin = margin(0, 0, 0, 0), 
    axis.text.x.bottom = element_blank(),  
    axis.ticks.x = element_blank(),
    strip.text.x.top = element_text(size=8), 
    strip.text.y.right = element_text(size=12),
    axis.text= element_text(size=8))+ 
  labs(x="SV gold standard", y="Tumour Purity", alpha="")

ggsave("figures/sv_heatmap_ggnested.png", height = 5.6, width = 16)
```

without facet_nested
```{r, fig.height=5.6, fig.width=15}
df_sv|>  mutate(ID=factor(ID,levels=id_levels), purity=factor(purity,levels=seq(10,100,10) |> as.character())) |> 
  ggplot(aes(x=ID,y=purity,fill=svtype, alpha=supp)) + 
  geom_tile(color="black") + 
  facet_grid2(tool~cellline, scale="free_x", space = "free_x", labeller = labeller(tool=public_tool_names)) +
  force_panelsizes(cols = c(3,15),rows = c(2,2,2,2), respect = TRUE) + 
  scale_alpha_manual(values = c(0.1, 1), labels = c("Missed", "Detected")) + 
  scale_fill_manual(values = sv_palette) + theme_bw() +
  theme(axis.text.x = element_text(angle=45,hjust=1),
    legend.position="none",
    legend.margin = margin(0, 0, 0, 0), 
    axis.text.x.bottom = element_blank(),  
    axis.ticks.x = element_blank(),
    strip.background = element_rect(fill="white"), 
    strip.text = element_text(size=12),
    axis.text= element_text(size=8))+ 
  labs(x="SV gold standard", y="Tumour purity", alpha="")

ggsave("figures/sv_heatmap_ggnested2.png", height = 5.6, width = 16)
```


```{r, cache=TRUE}
read_using_variantAnnotation <- function(filename) {
  tool <- str_extract(filename, c("delly","severus","savana","nanomonsv") |> paste(collapse = "|"))
  sample <- basename(filename) |> str_remove("\\.merged.vcf") |> str_split_i("\\.",i=1)
  readVcf(filename) |> info() |> as.data.frame() |> dplyr::select(SVTYPE,SUPP,SUPP_VEC) |> rownames_to_column("id") |> add_column(sample=sample, tool=tool)
}

df_sv_full <- list.files(paste0(workdir, "analysis/benchmark/svs"), pattern = "*\\.merged.vcf",recursive = TRUE, full.names = TRUE) |> map(read_using_variantAnnotation) |> list_rbind() |>  mutate(purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity), cellline=ifelse(startsWith(sample,'COLO829'),"COLO829","HCC1937"), svtype=ifelse(SVTYPE %in% c("INS","DUP"), "INS/DUP",SVTYPE), purity=factor(purity, levels=seq(10,100,10))) 
```

A general benchmarking

```{r, fig.height=6.6, fig.width=8}
temp_df_sv <- df_sv_full |> 
  group_by(SUPP_VEC, cellline, tool, purity) |> summarise(n=n(),total=sum(n)) |> 
  ungroup() |> group_by(cellline, tool, purity) |> 
  mutate(total=sum(n[SUPP_VEC!="01"]), total_called = sum(n[SUPP_VEC!="10"]), recall=n[SUPP_VEC=="11"]/total, precision=n[SUPP_VEC=="11"]/total_called) |> 
  mutate(recall_pos = -n[SUPP_VEC=="10"], precision_pos = total_called, recall=round(recall, digits = 4), precision = round(precision, digits = 4)) |> 
  ungroup() |> 
  dplyr::select(cellline,tool,purity,recall,precision, recall_pos, precision_pos) |> unique()

df_sv_full |> group_by(SUPP_VEC, cellline, tool, purity) |> summarise(n=n()) |> ungroup() |> mutate(value=ifelse(SUPP_VEC=="10",-n, n)) |> 
  ggplot(aes(x=value,y=purity)) + geom_col(aes(fill=SUPP_VEC)) + scale_fill_manual(values = mypal, labels=c("LR novel", "SR specific", "Concordant")) +
  geom_text(data = temp_df_sv,aes(x=recall_pos,label=recall*100), size=2, hjust=1.1,alpha=1) + 
  geom_text(data = temp_df_sv,aes(x=precision_pos,label=precision*100), size=2, hjust=-.1, alpha=1) + 
  facet_grid(tool~cellline, scales = "free_x",labeller = labeller(tool=public_tool_names)) +
  scale_x_continuous(expand = expansion(mult = c(0.1, 0.1))) +
  theme_bw(base_size = 12) + labs(x="Count",y="Tumour purity", fill="Category") + 
  theme(strip.background =element_rect(fill="white"), legend.position = "bottom", strip.text.y.right = element_text(size=12),
    axis.text= element_text(size=8))

ggsave("figures/sv_purity_benchmarking.png", height = 6.6, width = 8)
```

Currently, I think combining SVs detected by multiple tools is a good strategy to call as many novel SV as possibly while keep confidence. Here I compared the the intersect of four tools and pick any SVs that are detected by at least two tools with short-read 'gold standard'.

```{r, fig.height=2.5, fig.width=8}
read_using_variantAnnotation2 <- function(filename) {
  sample <- basename(filename) |> str_remove(".all_merged.vcf") |> str_remove("R10_sup_") |> str_split_i("\\.",i=1)
  readVcf(filename) |> info() |> as.data.frame() |> dplyr::select(SVTYPE,SUPP,SUPP_VEC) |> rownames_to_column("id") |> add_column(sample=sample)
}


df_sv_full_intersect <- list.files(paste0(workdir, "analysis/benchmark/svs"), pattern = "*.all_merged.vcf", full.names = TRUE) |>  map(read_using_variantAnnotation2) |> list_rbind() |>  mutate(purity=str_split_i(sample,"_",i=2), purity=ifelse(is.na(purity),"100",purity), cellline=ifelse(startsWith(sample,'COLO829'),"COLO829","HCC1937"), svtype=ifelse(SVTYPE %in% c("INS","DUP"), "INS/DUP",SVTYPE), purity=factor(purity, levels=seq(10,100,10))) 
  

temp_df_sv_intersect <- df_sv_full_intersect |> 
  group_by(SUPP_VEC, cellline, purity) |> summarise(n=n(),total=sum(n)) |> 
  ungroup() |> group_by(cellline,purity) |> 
  mutate(total=sum(n[SUPP_VEC!="01"]), total_called = sum(n[SUPP_VEC!="10"]), recall=n[SUPP_VEC=="11"]/total, precision=n[SUPP_VEC=="11"]/total_called) |> 
  mutate(recall_pos = -n[SUPP_VEC=="10"], precision_pos = total_called, recall=round(recall, digits = 4), precision = round(precision, digits = 4)) |> 
  ungroup() |> 
  dplyr::select(cellline,purity,recall,precision, recall_pos, precision_pos) |> unique()


df_sv_full_intersect |> group_by(SUPP_VEC, cellline, purity) |> summarise(n=n()) |> ungroup() |> mutate(value=ifelse(SUPP_VEC=="10",-n, n)) |> 
  ggplot(aes(x=value,y=purity)) + geom_col(aes(fill=SUPP_VEC)) + scale_fill_manual(values = mypal, labels=c("LR novel", "SR specific", "Concordant")) +
  geom_text(data = temp_df_sv_intersect,aes(x=recall_pos,label=recall*100), size=2, hjust=1.1,alpha=1) + 
  geom_text(data = temp_df_sv_intersect,aes(x=precision_pos,label=precision*100), size=2, hjust=-.1, alpha=1) + 
  facet_grid(~cellline, scales = "free_x") +
  scale_x_continuous(expand = expansion(mult = c(0.1, 0.1))) +
  theme_bw(base_size = 12) + labs(x="Count",y="Tumour purity", fill="Category") + 
  theme(strip.background =element_rect(fill="white"), legend.position = "bottom", strip.text.y.right = element_text(size=12),
    axis.text= element_text(size=8))

ggsave("figures/sv_intersect_purity_benchmarking.png",  height = 2.5, width = 8)
```

Seems it's not improve the recall, it means SVs that are not detected by Severus will not likely be detected by other tools we used as well.

## Pick short-read specific SVs that are not detected in all samples with all tools.

```{r, eval=FALSE, echo=TRUE}
df_sv |> filter(cellline=="COLO829") |> 
  group_by(`#CHROM`, POS, ID, svtype, supp) |> 
  summarise(n=n()) |> ungroup() |>
  complete(supp, ID, fill = list(n=0)) |>  filter(n==0, supp=="2") |> 
  mutate(id=sub("^[01]_", "", ID)) |> 
  dplyr::select(id) |>write_tsv("../../3.igv_check/sr_specific/colo829.sr_specific.id", col_names = FALSE)

df_sv |> filter(cellline=="HCC1937") |> filter(`#CHROM`!="chrX") |> 
  group_by(`#CHROM`, POS,ID, svtype, supp) |> 
  summarise(n=n()) |> ungroup() |>
  complete(supp, ID, fill = list(n=0)) |>  filter(n==0, supp=="2") |> 
  mutate(id=sub("^[01]_", "", ID)) |> 
  dplyr::select(id) |> write_tsv("../../3.igv_check/sr_specific/hcc1937.sr_specific.id", col_names = FALSE)
```

## Merge SVs called by different tool

```{r}
readVcf(paste0(workdir, "analysis/benchmark/svs/R10_sup_COLO829.COLO829_BL.tools_merged.vcf")) |> info() |> as.data.frame() |> 
  mutate(Delly=str_split_i(SUPP_VEC,"",1) |>  as.double(),
         nanomonsv=str_split_i(SUPP_VEC,"",2) |>  as.double(),
         SAVANA=str_split_i(SUPP_VEC,"",3) |>  as.double(),
         Severus=str_split_i(SUPP_VEC,"",4) |> as.double()) |> dplyr::select(Delly, nanomonsv, SAVANA, Severus) |> 
  upset(sets=c("Delly","nanomonsv","SAVANA","Severus"), keep.order = TRUE) |> ggplotify::as.ggplot()

ggsave("figures/sv_lr_tool_overlap_colo829.png",height = 5,width = 6)
```

```{r}
readVcf(paste0(workdir, "analysis/benchmark/svs/R10_sup_HCC1937.HCC1937_BL.tools_merged.vcf")) |> info() |> as.data.frame() |> 
  mutate(Delly=str_split_i(SUPP_VEC,"",1) |>  as.double(),
         nanomonsv=str_split_i(SUPP_VEC,"",2) |>  as.double(),
         SAVANA=str_split_i(SUPP_VEC,"",3) |>  as.double(),
         Severus=str_split_i(SUPP_VEC,"",4) |> as.double()) |> dplyr::select(Delly, nanomonsv, SAVANA, Severus) |> 
  upset(sets=c("Delly","nanomonsv","SAVANA","Severus"), keep.order = TRUE) |>  ggplotify::as.ggplot()

ggsave("figures/sv_lr_tool_overlap_hcc1937.png",height = 5,width = 6)
```





